src/frequenz/channels/utils/file_watcher.py

```diff
@@ -62,19 +64,26 @@ def __del__(self) -> None:
         """
         self._stop_event.set()
 
-    async def receive(self) -> Optional[pathlib.Path]:
+    async def _ready(self) -> None:
         """Wait for the next file event and return its path.
 
+        Raises:
+            StopAsyncIteration: When the channel is closed.
+
         Returns:
             Path of next file.
         """
-        while True:
-            changes = await self._awatch.__anext__()
-            for change in changes:
-                # Tuple of (Change, path) returned by watchfiles
-                if change is None or len(change) != 2:
-                    return None
-
-                _, path_str = change
-                path = pathlib.Path(path_str)
-                return path
+        # if there are messages waiting to be consumed, return immediately.
+        if self._changes:
+            return
+
+        self._changes = await self._awatch.__anext__()
+
+    def _get(self) -> pathlib.Path:
+        change = self._changes.pop()
+        # Tuple of (Change, path) returned by watchfiles
+        if change is None or len(change) != 2:
+            raise StopAsyncIteration()
+        _, path_str = change
+        path = pathlib.Path(path_str)
+        return path
```

Comment:
Oh, we are not even reporting what kind of change was observed? :thinking: :scream:

Reply:
I created #53 for this.